src/html-imports.js

```diff
@@ -9,718 +9,780 @@
+   *  synchronously if document is already done loading.
+   * @param {!function()} callback
+   */
+  function whenDocumentReady(callback) {
+    if (document.readyState !== 'loading') {
+      callback();
     } else {
-      element.addEventListener('load', markTargetLoaded);
-      element.addEventListener('error', markTargetLoaded);
+      document.addEventListener('readystatechange', function stateChanged() {
+        if (document.readyState !== 'loading') {
+          document.removeEventListener('readystatechange', stateChanged);
+          callback();
+        }
+      });
     }
   }
 
-  // make sure to catch any imports that are in the process of loading
-  // when this script is run.
-  (function() {
-    if (document.readyState === 'loading') {
-      var imports = document.querySelectorAll('link[rel=import]');
-      for (var i=0, l=imports.length, imp; (i<l) && (imp=imports[i]); i++) {
-        handleImport(imp);
+  /**
+   * Invokes the callback after all imports are loaded. Callback is called
+   * synchronously if imports are already done loading.
+   * @param {!function()} callback
+   */
+  function whenImportsReady(callback) {
+    let imports = /** @type {!NodeList<!HTMLLinkElement>} */
+      (document.querySelectorAll(rootImportsSelector));
+    const promises = [];
+    for (let i = 0, l = imports.length, imp; i < l && (imp = imports[i]); i++) {
+      if (!isElementLoaded(imp)) {
+        promises.push(whenElementLoaded(imp));
       }
     }
-  })();
-
-}
-
-// Fire the 'HTMLImportsLoaded' event when imports in document at load time
-// have loaded. This event is required to simulate the script blocking
-// behavior of native imports. A main document script that needs to be sure
-// imports have loaded should wait for this event.
-whenReady(function(detail) {
-  scope.ready = true;
-  scope.readyTime = new Date().getTime();
-  var evt = rootDocument.createEvent("CustomEvent");
-  evt.initCustomEvent("HTMLImportsLoaded", true, true, detail);
-  rootDocument.dispatchEvent(evt);
-});
-
-// Polyfill document.baseURI for browsers without it.
-if (!document.baseURI) {
-  var baseURIDescriptor = {
-    get: function() {
-      var base = document.querySelector('base');
-      return base ? base.href : window.location.href;
-    },
-    configurable: true
-  };
-
-  Object.defineProperty(document, 'baseURI', baseURIDescriptor);
-  Object.defineProperty(rootDocument, 'baseURI', baseURIDescriptor);
-}
+    if (promises.length) {
+      Promise.all(promises).then(() => callback());
+    } else {
+      callback();
+    }
+  }
 
-if (!useNative) {
-  function bootstrap() {
-    importer.bootDocument(document);
+  /**
+   * Returns the link that imported the element.
+   * @param {!Element} element
+   * @return {!HTMLLinkElement|undefined}
+   */
+  function importForElement(element) {
+    let target = element;
+    while ((target = target['__ownerImport'] || target.parentNode || target.host)) {
```

Comment:
The memoization here could be slightly cleaner.  Right now, even once you've memoized `__ownerImport`, you still go into the while loop and check if it's an import link.  Would be slightly better to fast-path if already memoized.  Pretty minor.